txnsync/encodedgroupsmarshalers.go

```diff
@@ -0,0 +1,826 @@
+			stub.Subsigs = make([][]crypto.MultisigSubsig, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskSubsigs.SetBit(i)
+		stub.Subsigs = append(stub.Subsigs, txn.Msig.Subsigs)
+	}
+}
+
+func (stub *txGroupsEncodingStub) finishDeconstructMsigs() {
+	stub.BitmaskVersion.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskThreshold.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskSubsigs.trimBitmask(int(stub.TotalTransactionsCount))
+}
+
+func (stub *txGroupsEncodingStub) deconstructLsigs(i int, txn transactions.SignedTxn) {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	if txn.Lsig.Logic != nil {
+		if len(stub.BitmaskLogic) == 0 {
+			stub.BitmaskLogic = make(bitmask, bitmaskLen)
+			stub.Logic = make([][]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskLogic.SetBit(i)
+		stub.Logic = append(stub.Logic, txn.Lsig.Logic)
+	}
+	if txn.Lsig.Args != nil {
+		if len(stub.BitmaskLogicArgs) == 0 {
+			stub.BitmaskLogicArgs = make(bitmask, bitmaskLen)
+			stub.LogicArgs = make([][][]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskLogicArgs.SetBit(i)
+		stub.LogicArgs = append(stub.LogicArgs, txn.Lsig.Args)
+	}
+	if !txn.Lsig.Sig.MsgIsZero() {
+		if len(stub.BitmaskSig) == 0 {
+			stub.BitmaskSig = make(bitmask, bitmaskLen)
+			stub.Sig = make([]byte, 0, int(stub.TotalTransactionsCount)*len(crypto.Signature{}))
+		}
+		stub.BitmaskSig.SetBit(i)
+		stub.Sig = append(stub.Sig, txn.Lsig.Sig[:]...)
+	}
+	if txn.Lsig.Msig.Version != 0 {
+		if len(stub.BitmaskVersion) == 0 {
+			stub.BitmaskVersion = make(bitmask, bitmaskLen)
+			stub.Version = make([]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskVersion.SetBit(i)
+		stub.Version = append(stub.Version, txn.Lsig.Msig.Version)
+	}
+	if txn.Lsig.Msig.Threshold != 0 {
+		if len(stub.BitmaskThreshold) == 0 {
+			stub.BitmaskThreshold = make(bitmask, bitmaskLen)
+			stub.Threshold = make([]byte, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskThreshold.SetBit(i)
+		stub.Threshold = append(stub.Threshold, txn.Lsig.Msig.Threshold)
+	}
+	if txn.Lsig.Msig.Subsigs != nil {
+		if len(stub.BitmaskSubsigs) == 0 {
+			stub.BitmaskSubsigs = make(bitmask, bitmaskLen)
+			stub.Subsigs = make([][]crypto.MultisigSubsig, 0, int(stub.TotalTransactionsCount))
+		}
+		stub.BitmaskSubsigs.SetBit(i)
+		stub.Subsigs = append(stub.Subsigs, txn.Lsig.Msig.Subsigs)
+	}
+}
+
+func (stub *txGroupsEncodingStub) finishDeconstructLsigs() {
+	stub.BitmaskLogic.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskLogicArgs.trimBitmask(int(stub.TotalTransactionsCount))
+}
+
+func (stub *txGroupsEncodingStub) deconstructTransactions(i int, txn transactions.SignedTxn) error {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	txTypeByte, err := TxTypeToByte(txn.Txn.Type)
+	if err != nil {
+		return fmt.Errorf("failed to deconstructTransactions: %v", err)
+	}
+	if txTypeByte != 0 {
```

Comment:
two things - 
1. avoid comparing to magic numbers.
2. 0 seems to be a payment. How come it didn't break any of the unit testing ?

Reply:
So the way I did it was to first decompose it such that payment txns were omitted via the bitmask. Later I checked if a different txtype were more common, and if so, I recomputed bitmask / txtype field to fit the other txtype. So the result produced is still correct.

But now that you mention it, I can make this code more clean by not assuming everything is a payment txn at the beginning.